### PR TITLE
[ci][docs] fix link checking action by switching from linkchecker to lychee and update some links

### PR DESCRIPTION
**Pull request by StrikerRUS**
All our recent link check jobs are failing: https://github.com/microsoft/LightGBM/actions/workflows/linkchecker.yml. Last successful job was 5 months ago.

The most errors are about 429 Rate Limit GitHub urls, but others are quite random ones. It looks like our current `linkchecker` isn't reliable enough.

Initially I thought to use [`linkinator`](https://github.com/JustinBeckwith/linkinator) but then found [`lychee`](https://lychee.cli.rs/). This tool looks quite simple to use but [powerful](https://lychee.cli.rs/comparison/).

Latest `linkchecker` results:

```
That's it. 1058 links in 1076 URLs checked. 3 warnings found. 4 errors found.
Stopped checking at 2025-09-11 08:54:22+000 (48 minutes, 31 seconds)
```

New `lychee` results:

```
🔍 2188 Total (in 27s) ✅ 2167 OK 🚫 0 Errors 👻 21 Excluded
```

I left some inline comments below to help understand some non obvious moments.

### CODE REVIEW

File: .ci/test-docs.sh [hunk 1]
```diff
@@ -15,6 +15,17 @@ make -C docs html || exit 1
...
     # check docs for broken links
-    pip install 'linkchecker>=10.5.0'
-    linkchecker --config=./docs/.linkcheckerrc ./docs/_build/html/*.html || exit 1
+    conda install -y -n test-env 'lychee>=0.20.1'
+    # to see all gained files add "--dump-inputs" flag
+    # to see all gained links add "--dump" flag
+    lychee \
+        "--config=./docs/.lychee.toml" \
+        "--exclude-path=(^|/)docs/.*\.rst" \
+        "--github-token=${SECRETS_WORKFLOW}" \
```
Comment (StrikerRUS) [inline .ci/test-docs.sh]: This helps to overcome 429 Rate Limit errors from GitHub greatly: https://lychee.cli.rs/troubleshooting/rate-limits/#github-rate-limiting

File: .ci/test-docs.sh [hunk 2]
```diff
@@ -15,6 +15,17 @@ make -C docs html || exit 1
...
 if [[ $TASK == "check-links" ]]; then
     # check docs for broken links
-    pip install 'linkchecker>=10.5.0'
-    linkchecker --config=./docs/.linkcheckerrc ./docs/_build/html/*.html || exit 1
+    conda install -y -n test-env 'lychee>=0.20.1'
+    # to see all gained files add "--dump-inputs" flag
+    # to see all gained links add "--dump" flag
+    lychee \
+        "--config=./docs/.lychee.toml" \
+        "--exclude-path=(^|/)docs/.*\.rst" \
```
Comment (StrikerRUS) [inline .ci/test-docs.sh]: We will check only final `.html` files in `docs`.

File: R-package/README.md [hunk 3]
```diff
@@ -20,7 +20,6 @@
     - [Code Coverage](#code-coverage)
 * [Updating Documentation](#updating-documentation)
 * [Preparing a CRAN Package](#preparing-a-cran-package)
-* [External Repositories](#external-unofficial-repositories)
```
Comment (StrikerRUS) [inline R-package/README.md:23]: This section was removed sometime ago.

File: R-package/README.md [hunk 4]
```diff
@@ -77,7 +76,7 @@ CXX=g++-8
 CXX11=g++-8
 ```
 
-### Installing from Source with CMake <a name="install"></a>
```
Comment (StrikerRUS) [inline R-package/README.md:80]: [Fix](https://github.com/lycheeverse/lychee/pull/1839) has been already merged but let's not wait for a new release with that fix.

File: R-package/cran-comments.md [hunk 5]
```diff
@@ -706,7 +706,7 @@ All authors from the original LightGBM paper have been added to Authors@R as `"a
 
 lightgbm's code does use `<<-`, but it does not modify the global environment. The uses of `<<-` in R/lgb.interprete.R and R/callback.R are in functions which are called in an environment created by the lightgbm functions that call them, and this operator is used to reach one level up into the calling function's environment.
 
-We chose to wrap our examples in `\donttest{}` because we found, through testing on https://builder.r-hub.io/ and in our own continuous integration environments, that their run time varies a lot between platforms, and we cannot guarantee that all examples will run in under 5 seconds. We intentionally chose `\donttest{}` over `\donttest{}` because this item in the R 4.0.0 changelog (https://cran.r-project.org/doc/manuals/r-devel/NEWS.html) seems to indicate that \donttest will be ignored by CRAN's automated checks:
```
Comment (StrikerRUS) [inline R-package/cran-comments.md:709]: This project isn't maintained anymore: https://github.com/r-hub/r-builder

```
Public archive
```

File: docs/.lychee.toml [hunk 6]
```diff
@@ -0,0 +1,27 @@
+verbose = "info"
```
Comment (StrikerRUS) [inline docs/.lychee.toml:1]: Options description: https://lychee.cli.rs/usage/config/

File: docs/.lychee.toml [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+no_progress = false
+cache = false
+scheme = ["http", "https", "file"]
+include_mail = false
+include_fragments = true
```
Comment (StrikerRUS) [inline docs/.lychee.toml:6]: "fragments" means anchors.

File: docs/.lychee.toml [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+no_ignore = true
+insecure = false
+require_https = true
```
Comment (StrikerRUS) [inline docs/.lychee.toml:9]: Nice option that requires to write secure `https://` links where they are available.

File: docs/.lychee.toml [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+accept = ["100..=103", "200..=299"]
+user_agent = "curl/7.88.1"
+header = {"User-Agent" = "curl/7.88.1"}
```
Comment (StrikerRUS) [inline docs/.lychee.toml:12]: Without these settings some sites block lychee and return 403 code.

File: docs/.lychee.toml [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+timeout = 30
+retry_wait_time = 10
+max_concurrency = 10
+# remove anchors from GitHub URLs to overcome https://github.com/lycheeverse/lychee/issues/1729
+remap = [
+    '(?P<host>^https://github\.com)/(?P<path>.*)#(?P<anchor>.*)$ $host/$path/',
+]
```
Comment (StrikerRUS) [inline docs/.lychee.toml:19]: We've already faced this problem, e.g. here #6562.

File: docs/.lychee.toml [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+exclude = [
+    '^https://www\.swig\.org/download\.html$',
```
Comment (StrikerRUS) [inline docs/.lychee.toml:21]: Aggressive protection from bots.

File: docs/.lychee.toml [hunk 12]
```diff
@@ -0,0 +1,27 @@
...
+    '^https://.*\.n[eur]*ips\.cc/.*',
```
Comment (StrikerRUS) [inline docs/.lychee.toml]: Hope this is temporarily because NIPS sites went down just two days ago.

File: docs/.lychee.toml [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+    '^https://www\.amd\.com/en/support\.html$',
```
Comment (StrikerRUS) [inline docs/.lychee.toml:23]: Our very old friend: #3530.
It can be reached successfully only with **empty** user agent. But with such configuration many other sites return 403.

File: docs/.lychee.toml [hunk 14]
```diff
@@ -0,0 +1,27 @@
...
+    '^https://www\.jstor\.org/stable/2281952$',
+    '^https://dl\.acm\.org/doi/10\.1145/3298689\.3347033$',
```
Comment (StrikerRUS) [inline docs/.lychee.toml:25]: Aggressive protection from bots.

File: docs/.lychee.toml [hunk 15]
```diff
@@ -0,0 +1,27 @@
...
+    '^https://packages\.ubuntu\.com/search.*',
```
Comment (StrikerRUS) [inline docs/.lychee.toml:26]: Flaky: sometimes return 500 code.

File: docs/Experiments.rst [hunk 16]
```diff
@@ -23,7 +23,7 @@ We used 5 datasets to conduct our comparison experiments. Details of data are li
 +===========+=======================+=================================================================================+=============+==========+==============================================+
 | Higgs     | Binary classification | `link <https://archive.ics.uci.edu/dataset/280/higgs>`__                        | 10,500,000  | 28       | last 500,000 samples were used as test set   |
 +-----------+-----------------------+---------------------------------------------------------------------------------+-------------+----------+----------------------------------------------+
-| Yahoo LTR | Learning to rank      | `link <https://webscope.sandbox.yahoo.com/catalog.php?datatype=c>`__            | 473,134     | 700      | set1.train as train, set1.test as test       |
```
Comment (StrikerRUS) [inline docs/Experiments.rst:26]: No longer exists. And using wayback machine doesn't allow to download the dataset due to checkouting mechanism.

File: docs/Parameters.rst [hunk 17]
```diff
@@ -1265,7 +1265,7 @@ Metric Parameters
 
       -  ``binary_error``, for one sample: ``0`` for correct classification, ``1`` for error classification
 
-      -  ``auc_mu``, `AUC-mu <http://proceedings.mlr.press/v97/kleiman19a/kleiman19a.pdf>`__
+      -  ``auc_mu``, `AUC-mu <https://proceedings.mlr.press/v97/kleiman19a.html>`__
```
Comment (StrikerRUS) [inline docs/Parameters.rst:1268]: Why direct pdf links are bad: https://github.com/microsoft/LightGBM/pull/5923#discussion_r1720750667

File: examples/README.md [hunk 18]
```diff
@@ -13,50 +13,50 @@ Machine Learning Challenge Winning Solutions
...
+|3rd      | [M5 Forecasting - Uncertainty](https://www.kaggle.com/c/m5-forecasting-uncertainty) | [link](https://www.kaggle.com/competitions/m5-forecasting-uncertainty/writeups/ouranos-3rd-place-solution) | 2020.7 |
+|3rd      | [ALASKA2 Image Steganalysis](https://www.kaggle.com/c/alaska2-image-steganalysis) | [link](https://www.kaggle.com/competitions/alaska2-image-steganalysis/writeups/kaizaburochubachi-3rd-place-solution) | 2020.7 |
+|1st      | [M5 Forecasting - Accuracy](https://www.kaggle.com/c/m5-forecasting-accuracy) | [link](https://www.kaggle.com/competitions/m5-forecasting-accuracy/writeups/yeonjun-in-stu-1st-place-solution) | 2020.6 |
+|2nd      | [COVID19 Global Forecasting (Week 5)](https://www.kaggle.com/c/covid19-global-forecasting-week-5) | [link](https://www.kaggle.com/competitions/covid19-global-forecasting-week-5/writeups/kaz-some-ml-a-lot-of-judgement-and-luck) | 2020.5 |
 |3rd      | [COVID19 Global Forecasting (Week 5)](https://www.kaggle.com/c/covid19-global-forecasting-week-5) | [link](https://www.kaggle.com/c/covid19-global-forecasting-week-5/discussion/143029) | 2020.5 |
```
Comment (StrikerRUS) [inline examples/README.md]: Not changed Kaggle links have been deleted
But related comments are still accessible and probably can be useful for someone.

image

File: examples/README.md [hunk 19]
```diff
@@ -13,50 +13,50 @@ Machine Learning Challenge Winning Solutions
...
+| 1st     | [Santander Customer Transaction Prediction](https://www.kaggle.com/c/santander-customer-transaction-prediction) | [link](https://www.kaggle.com/competitions/santander-customer-transaction-prediction/writeups/wizardry-1-solution) | 2019.4 |
+| 2nd     | [Santander Customer Transaction Prediction](https://www.kaggle.com/c/santander-customer-transaction-prediction) | [link](https://www.kaggle.com/competitions/santander-customer-transaction-prediction/writeups/2nd-place-solution) | 2019.4 |
+| 3rd     | [Santander Customer Transaction Prediction](https://www.kaggle.com/c/santander-customer-transaction-prediction) | [link](https://www.kaggle.com/competitions/santander-customer-transaction-prediction/writeups/rock-physics-science-3rd-place-solution-summary-an) | 2019.4 |
+| 1st     | [PetFinder.my Adoption Prediction](https://www.kaggle.com/c/petfinder-adoption-prediction) | [link](https://www.kaggle.com/competitions/petfinder-adoption-prediction/writeups/kaggler-ja-wodori-1st-place-solution-summary) | 2019.4 |
```
Comment (StrikerRUS) [inline examples/README.md:43]: [PetFinder.my Contest: 1st Place Winner Disqualified](https://www.kaggle.com/c/petfinder-adoption-prediction/discussion/125436)

File: .ci/test-docs.sh (hunk 1 above)
Comment (jameslamb) [reply .ci/test-docs.sh]: This makes sense, but I think we should stop using that secret (which I think is tied to someone's personal account, maybe @shiyu1994 's) and start using the short-term one that GitHub provisions automatically for the repo.

I'm planning to try that for the valgrind job in #7012

It's described here:

https://docs.github.com/en/actions/reference/workflows-and-actions/contexts#github-context

You could set it like this in the GitHub Actions YAML:

```yaml
env:
  GITHUB_TOKEN: ${{ github.token }}
```

Benefits of this:

* short-lived (so limited impact if it's accidentally leaked in logs or something)
* tied to the repo, not an individual (so no need to ask any one maintainer to update it if/when it expires)

File: docs/.lychee.toml (hunk 13 above)
Comment (jameslamb) [reply docs/.lychee.toml:23]: > _our very old friend_

ah it's been a while 😆

File: examples/README.md (hunk 19 above)
Comment (jameslamb) [reply examples/README.md:43]: WOW had not heard about this!

File: .ci/test-docs.sh (hunk 2 above)
Comment (jameslamb) [reply .ci/test-docs.sh]: I think we should also add this to `lychee.toml`: https://lychee.cli.rs/recipes/excluding-paths/

The more that gets put into that file, the easier it'd be to run this interactively during development.

File: .ci/test-docs.sh (hunk 1 above)
Comment (StrikerRUS) [reply .ci/test-docs.sh]: OK, I'll replace it with default token.

File: .ci/test-docs.sh (hunk 2 above)
Comment (StrikerRUS) [reply .ci/test-docs.sh]: My initial thoughts were to group all files-related options into arguments, and all other options place into `.lychee.toml` (`GITHUB_SECRET` is an exclusion because you cannot place env variable into TOML). But I don't have a strong preference.

**Comment by StrikerRUS** [comment]
@jameslamb Thanks a lot for reviewing this! Nice to hear that you are already familiar with `lychee` 😃.

Would it be OK for you if I push changes for GitHub token and paths exclusion in this PR, and move this to GitHub Action in a follow-up PR?

**Comment by jameslamb** [comment]
> _Would it be OK for you if I push changes for GitHub token and paths exclusion in this PR, and move this to GitHub Action in a follow-up PR?_

Yes absolutely!

**Comment by StrikerRUS** [comment]
@jameslamb I've pushed changes I promised you in https://github.com/microsoft/LightGBM/pull/7027#issuecomment-3284550847. Could you please check them when have time? Please notice I have to add stackoverflow site to ignore list due to the following errors:

```
Issues found in 3 inputs. Find details below.

[docs/_build/html/FAQ.html]:
[403] https://stackoverflow.com/questions/18085571/pip-install-error-setup-script-specifies-an-absolute-path | Rejected status code (this depends on your "accept" configuration): Forbidden

[docs/_build/html/Parameters-Tuning.html]:
[403] https://stats.stackexchange.com/questions/317073/explanation-of-min-child-weight-in-xgboost-algorithm | Rejected status code (this depends on your "accept" configuration): Forbidden

[README.md]:
[403] https://stackoverflow.com/questions/ask?tags=lightgbm | Rejected status code (this depends on your "accept" configuration): Forbidden
[403] https://stackoverflow.com/questions/tagged/lightgbm?sort=votes | Rejected status code (this depends on your "accept" configuration): Forbidden

🔍 2208 Total (in 26s) ✅ 2183 OK 🚫 4 Errors 👻 21 Excluded
```

Now results look like the following:

```
🔍 2208 Total (in 26s) ✅ 2183 OK 🚫 0 Errors 👻 25 Excluded
```

File: docs/.lychee.toml (hunk 8 above)
Comment (jameslamb) [reply docs/.lychee.toml:9]: love this

File: .ci/test-docs.sh (hunk 1 above)
Comment (jameslamb) [reply .ci/test-docs.sh]: thanks!

File: .ci/test-docs.sh (hunk 2 above)
Comment (jameslamb) [reply .ci/test-docs.sh]: Thanks for moving the `--exclude-path` into `.lychee.toml`, I don't feel TOO strongly about it either but I do like the current state of this branch where most of the configuration is in that file instead of `test-docs.sh`.

**Comment by StrikerRUS** [comment]
> ... I'm happy to see this only takes around 2 minutes to run ([build link](https://github.com/microsoft/LightGBM/actions/runs/18198140129)) compared to nearly an hour for the `linkcheck`-based job ([build link](https://github.com/microsoft/LightGBM/actions/runs/18400419813/job/52428172325))!!!

Yeah, thanks to GitHub token lychee mechanism we don't need to manually slow down link checking to avoid 429 Rate limit errors.
